### ISSUE TITLE
uoe: occasional spurious joint angle measurement

### ISSUE THREAD

**Issue by mauricefallon** [NASA-JSC-Robotics/valkyrie#17]
We have noticed very occasional spikes in the controller behaviour which we have now pinned down to the joint position reported from NASA containing a corrupted reading.

In this short log for a single tic the hip joint reads a joint angle of zero:
http://terminator.robots.inf.ed.ac.uk/public/scs/20160314_170257/20160314_170257_ValkyrieRosControlController_UOEPossibleMissedTickFromNASA.tar.gz

**Comment by mauricefallon** [comment]
Tic 1:
![tic1](https://cloud.githubusercontent.com/assets/6404227/13755186/e3ec8dd2-ea11-11e5-9ca0-2fcb24da8750.png)

Tic 2:
![tic2](https://cloud.githubusercontent.com/assets/6404227/13755189/e78f0eb0-ea11-11e5-8494-b2540d318d24.png)

Tic 3:
![tic3](https://cloud.githubusercontent.com/assets/6404227/13755191/eaa7f198-ea11-11e5-8f4b-78de975dbfe5.png)

What could be the source of this issue?

Note: that we logged the value both in IHMC's log and also in our log of the raw ROS messages from NASA.

**Comment by jlack1987** [comment]
Nice catch @mauricefallon, have you caught it more than once and if you have is it always in the same joint that you see the spike?

**Comment by jlack1987** [comment]
talked a bit with @jjholley and @MathisFrankB and we have a few questions. When you say you have a "log of the raw ROS messages from NASA", do you mean like a rosbag of the /joint_states topic that we publish? If so do you see it happen there also?

**Comment by mauricefallon** [comment]
We have noticed it a few times before. It never causes a fall but it causes the robot to screech and sway. I'll send more logs when we observe it again. (difficult to find in the current logs)

Yes, I see it on the /joint_states topic. (Actually its /joint_states logged in an LCM log)
So I am confident that its unrelated to IHMC.

**Comment by jlack1987** [comment]
ok cool, if it's seen in the /joint_states that we publish then it's definitely coming from us and not IHMC as long as the /joint_states your logging is coming from us which I assume it is. 

Next time you see this could you turn off logic power which will cause our logger to dump a log file to the NAS to /media/val_smt_logs. Could you send that to us?

**Comment by jlack1987** [comment]
Alright @mauricefallon, we have finally observed this one in one of our own data logs here at JSC on another robot. This just means the issue is definitely on our side. We have a fix for it but want to do a bit more investigating to make sure we put the fix at the proper level. We'll let you know when we have pushed a fix in code.

**Comment by mauricefallon** [comment]
sure, BTW, we have seen it regularly since, but haven't bothered to isolate the snippet. I assume you dont need further evidence.

**Comment by jlack1987** [comment]
Yeah we have everything we need data wise. We are going to do some poking around to figure out exactly who is the culprit(synapse driver, turbodriver, API, etc), but we have everything we need and will get a fix out some time soon. Keep in mind though that we are delivering more robots soon, so it might be a few weeks before we get a chance to run all the tests we have planned.

**Comment by jlack1987** [comment]
Just pushed a fix to this yesterday. It looks like the cause lies somewhere in our code for the synapse driver. We have some work going on to refactor all of our synapse code, so the true fix to this will come later. For now we have implemented a glitch filter that will remove the glitches. We did some limited testing and found that it works pretty well.

A few things to keep in mind. When a glitch happens you will miss a single tic on that particular joint's position measurement. If a glitch happens 2 tic's in a row then the filter will fail and it won't get properly filtered out. Also if the glitch happens on a joint and that joint's position reading is below the glitch filter's threshold, those also won't get filtered out. All that being said in the testing we did over the weekend we seem to successfully filter out the vast majority of all glitches that happen.
